File: frontend/android/app/src/main/java/com/quietcue/app/domain/AlertModels.kt
package com.quietcue.app.domain

data class DetectedAlert(
    val eventId: String,
    val event: String,
    val confidence: Float,
    val category: String,
    val pattern: String,
    val profileName: String,
    val sourceLabel: String,
    val totalLatencyMs: Int,
    val requiresAcknowledgement: Boolean,
    val simulated: Boolean,
    val fallbackToPhone: Boolean,
    val hapticActive: Boolean = false,
    val acknowledgedAtMs: Long? = null,
) {
    val displayName: String
        get() = if (event.startsWith("custom:") && sourceLabel.startsWith("enrolled: ")) {
            sourceLabel.removePrefix("enrolled: ")
        } else event
            .split('_')
            .joinToString(" ") { word -> word.replaceFirstChar(Char::uppercase) }
}

enum class InferenceDevice(val wireValue: String, val displayName: String) {
    COPILOT_PC("copilot_pc", "PC"),
    SAMSUNG_PHONE("samsung_phone", "Samsung"),
    ;

    companion object {
        fun fromWire(value: String?): InferenceDevice = entries.firstOrNull {
            it.wireValue == value
        } ?: COPILOT_PC
    }
}

data class RuntimeState(
    val backendConnected: Boolean = false,
    val audioSourceConnected: Boolean = false,
    val backendProfileName: String? = null,
    val latestAlert: DetectedAlert? = null,
    val errorMessage: String? = null,
    val stopInProgress: Boolean = false,
    val requestedInferenceDevice: InferenceDevice = InferenceDevice.COPILOT_PC,
    val activeInferenceDevice: InferenceDevice = InferenceDevice.COPILOT_PC,
    val samsungInferenceAvailable: Boolean = false,
    val inferenceSwitchPending: Boolean = false,
    val inferenceRoutingError: String? = null,
)

File: frontend/ios/QuietCue/Domain/AlertModels.swift
import Foundation

struct DetectedAlert: Equatable, Hashable {
    var eventId: String
    var event: String
    var confidence: Float
    var category: String
    var pattern: String
    var profileName: String
    var sourceLabel: String
    var totalLatencyMs: Int
    var requiresAcknowledgement: Bool
    var simulated: Bool
    var fallbackToPhone: Bool
    var hapticActive: Bool = false
    var acknowledgedAtMs: Int64? = nil

    var displayName: String {
        if event.hasPrefix("custom:"), sourceLabel.hasPrefix("enrolled: ") {
            return String(sourceLabel.dropFirst("enrolled: ".count))
        }
        return event.split(separator: "_").map { $0.prefix(1).uppercased() + $0.dropFirst() }.joined(separator: " ")
    }
}

enum InferenceDevice: String, CaseIterable {
    case copilotPC = "copilot_pc"
    case phone = "samsung_phone"

    var displayName: String {
        switch self {
        case .copilotPC: return "PC"
        case .phone: return "iPhone"
        }
    }

    static func fromWire(_ value: String?) -> InferenceDevice {
        allCases.first { $0.rawValue == value } ?? .copilotPC
    }
}

struct RuntimeState: Equatable {
    var backendConnected: Bool = false
    var audioSourceConnected: Bool = false
    var backendProfileName: String? = nil
    var latestAlert: DetectedAlert? = nil
    var errorMessage: String? = nil
    var stopInProgress: Bool = false
    var requestedInferenceDevice: InferenceDevice = .copilotPC
    var activeInferenceDevice: InferenceDevice = .copilotPC
    var phoneInferenceAvailable: Bool = false
    var inferenceSwitchPending: Bool = false
    var inferenceRoutingError: String? = nil
}

/// Connection state of the phone's own microphone stream to the hub (the
/// role the Uno Q plays in the hardware demo).
struct EdgeStreamState: Equatable {
    enum Phase: Equatable {
        case idle
        case connecting
        case streaming
        case retrying(seconds: Int)
        case failed(String)
    }

    var phase: Phase = .idle
    var microphoneAuthorized: Bool = false
    var chunksSent: Int = 0
    var lastRmsDbfs: Float? = nil
    var lastRoundTripMs: Int? = nil
    var hubNodeId: String? = nil
    var hapticsDelivered: Int = 0
    var lastHapticPattern: String? = nil
    var hapticPlaying: Bool = false

    var isStreaming: Bool { phase == .streaming }
}
